ethereum/permissioning/src/test/java/tech/pegasys/pantheon/ethereum/permissioning/node/NodePermissioningControllerFactoryTest.java

```diff
@@ -0,0 +1,100 @@
+package tech.pegasys.pantheon.ethereum.permissioning.node;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import tech.pegasys.pantheon.ethereum.core.Address;
+import tech.pegasys.pantheon.ethereum.core.Synchronizer;
+import tech.pegasys.pantheon.ethereum.permissioning.LocalPermissioningConfiguration;
+import tech.pegasys.pantheon.ethereum.permissioning.NodePermissioningControllerFactory;
+import tech.pegasys.pantheon.ethereum.permissioning.PermissioningConfiguration;
+import tech.pegasys.pantheon.ethereum.permissioning.SmartContractPermissioningConfiguration;
+import tech.pegasys.pantheon.ethereum.transaction.TransactionSimulator;
+import tech.pegasys.pantheon.util.enode.EnodeURL;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class NodePermissioningControllerFactoryTest {
+
+  @Mock private Synchronizer synchronizer;
+  @Mock private TransactionSimulator transactionSimulator;
+
+  Collection<EnodeURL> bootnodes = Collections.emptyList();
+  LocalPermissioningConfiguration localPermissioningConfig;
+  SmartContractPermissioningConfiguration smartContractPermissioningConfiguration;
+  PermissioningConfiguration config;
+
+  @Before
+  public void setUp() {
+    localPermissioningConfig = LocalPermissioningConfiguration.createDefault();
+    smartContractPermissioningConfiguration = new SmartContractPermissioningConfiguration();
+    config =
+        new PermissioningConfiguration(
+            Optional.of(localPermissioningConfig),
+            Optional.of(smartContractPermissioningConfiguration));
+  }
+
+  @Test
+  public void testCreateWithNeitherPermissioningEnabled() {
+    NodePermissioningControllerFactory factory = new NodePermissioningControllerFactory();
+    NodePermissioningController controller =
+        factory.create(config, synchronizer, bootnodes, transactionSimulator);
+
+    List<NodePermissioningProvider> providers = controller.getProviders();
+    assertThat(providers.size()).isEqualTo(0);
+  }
+
+  @Test
+  public void testCreateWithSmartContractNodePermissioningEnabledOnly() {
+    smartContractPermissioningConfiguration.setSmartContractAddress(
+        Address.fromHexString("0x0000000000000000000000000000000000001234"));
+    smartContractPermissioningConfiguration.setSmartContractNodeWhitelistEnabled(true);
+
+    NodePermissioningControllerFactory factory = new NodePermissioningControllerFactory();
+    NodePermissioningController controller =
+        factory.create(config, synchronizer, bootnodes, transactionSimulator);
+
+    List<NodePermissioningProvider> providers = controller.getProviders();
+    assertThat(providers.size()).isEqualTo(1);
+
+    NodePermissioningProvider p1 = providers.get(0);
+    assertThat(p1).isNotNull();
```

Comment:
Instead of checking that it is not null, we should probably check that we have the right type:
```
assertThat(p1).isInstanceOf(NodeSmartContractPermissioningController.class);
```

Reply:
done